File: ruby/lib/resp_bench/config/phase_config.rb
# frozen_string_literal: true

module RespBench
  module Config
    # Configuration for a benchmark phase.
    class PhaseConfig
      DEFAULT_PIPELINE_DEPTH = 1
      DEFAULT_WARMUP_REQUESTS = 1

      attr_accessor :id, :description, :connections, :cps_limit, :rps_limit,
                    :pipeline_depth, :warmup_requests, :completion, :keyspace, :commands

      def initialize(
        id:,
        connections:,
        completion:,
        keyspace:,
        commands:,
        description: nil,
        cps_limit: -1,
        rps_limit: -1,
        pipeline_depth: DEFAULT_PIPELINE_DEPTH,
        warmup_requests: DEFAULT_WARMUP_REQUESTS
      )
        @id = id
        @description = description
        @connections = connections
        @cps_limit = cps_limit || -1
        @rps_limit = rps_limit || -1
        @pipeline_depth = pipeline_depth || DEFAULT_PIPELINE_DEPTH
        @warmup_requests = warmup_requests || DEFAULT_WARMUP_REQUESTS
        @completion = completion
        @keyspace = keyspace
        @commands = commands
      end

      def cps_limit?
        @cps_limit.positive?
      end

      def rps_limit?
        @rps_limit.positive?
      end

      def effective_pipeline_depth
        @pipeline_depth.positive? ? @pipeline_depth : DEFAULT_PIPELINE_DEPTH
      end

      def to_h
        {
          id: @id,
          description: @description,
          connections: @connections,
          cps_limit: @cps_limit,
          rps_limit: @rps_limit,
          pipeline_depth: @pipeline_depth,
          warmup_requests: @warmup_requests,
          completion: @completion.to_h,
          keyspace: @keyspace.to_h,
          commands: @commands.map(&:to_h)
        }
      end
    end
  end
end

File: python/src/resp_bench/config/phase_config.py
"""Configuration for a single benchmark phase."""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from .command_config import CommandConfig
from .completion_config import CompletionConfig
from .keyspace_config import KeyspaceConfig

DEFAULT_PIPELINE_DEPTH = 1
DEFAULT_WARMUP_REQUESTS = 1


@dataclass
class PhaseConfig:
    id: str
    connections: int
    completion: CompletionConfig
    keyspace: KeyspaceConfig
    commands: List[CommandConfig]
    description: Optional[str] = None
    cps_limit: int = -1
    rps_limit: int = -1
    pipeline_depth: int = DEFAULT_PIPELINE_DEPTH
    warmup_requests: int = DEFAULT_WARMUP_REQUESTS

    def __post_init__(self) -> None:
        if self.cps_limit is None:
            self.cps_limit = -1
        if self.rps_limit is None:
            self.rps_limit = -1
        if self.pipeline_depth is None:
            self.pipeline_depth = DEFAULT_PIPELINE_DEPTH
        if self.warmup_requests is None:
            self.warmup_requests = DEFAULT_WARMUP_REQUESTS

    def has_cps_limit(self) -> bool:
        return self.cps_limit > 0

    def has_rps_limit(self) -> bool:
        return self.rps_limit > 0

    def effective_pipeline_depth(self) -> int:
        return self.pipeline_depth if self.pipeline_depth > 0 else DEFAULT_PIPELINE_DEPTH
